sdks/java/extensions/sql/datacatalog/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datacatalog/SchemaUtils.java

```diff
@@ -0,0 +1,98 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.meta.provider.datacatalog;
+
+import static org.apache.beam.sdk.schemas.Schema.toSchema;
+
+import com.google.cloud.datacatalog.ColumnSchema;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.extensions.sql.impl.utils.CalciteUtils;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.schemas.Schema.Field;
+import org.apache.beam.sdk.schemas.Schema.FieldType;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.collect.ImmutableMap;
+
+class SchemaUtils {
+
+  private static final Map<String, FieldType> FIELD_TYPES =
+      ImmutableMap.<String, FieldType>builder()
+          .put("TYPE_BOOL", FieldType.BOOLEAN)
+          .put("TYPE_BYTES", FieldType.BYTES)
+          .put("TYPE_DATE", FieldType.logicalType(new CalciteUtils.DateType()))
+          .put("TYPE_DATETIME", FieldType.DATETIME)
+          .put("TYPE_DOUBLE", FieldType.DOUBLE)
+          .put("TYPE_FLOAT", FieldType.DOUBLE)
+          .put("TYPE_INT32", FieldType.INT32)
+          .put("TYPE_INT64", FieldType.INT64)
+          .put("TYPE_STRING", FieldType.STRING)
+          .put("TYPE_TIME", FieldType.logicalType(new CalciteUtils.TimeType()))
+          .put("TYPE_TIMESTAMP", FieldType.DATETIME)
+          .put(
+              "TYPE_MAP<TYPE_STRING, TYPE_STRING>",
+              FieldType.map(FieldType.STRING, FieldType.STRING))
+          .build();
+
+  /** Convert DataCatalog schema to Beam schema. */
+  static Schema fromDataCatalog(com.google.cloud.datacatalog.Schema dcSchema) {
+    return fromColumnsList(dcSchema.getColumnsList());
+  }
+
+  private static Schema fromColumnsList(List<ColumnSchema> columnsMap) {
+    return columnsMap.stream().map(SchemaUtils::toBeamField).collect(toSchema());
+  }
+
+  private static Field toBeamField(ColumnSchema column) {
+    String name = column.getColumn();
+
+    // basic field type
+    FieldType fieldType = getBeamFieldType(column);
+    Field field = Field.of(name, fieldType);
+
+    // set the nullable flag, or convert to a list if repeated
+    if (Strings.isNullOrEmpty(column.getMode()) || "NULLABLE".equals(column.getMode())) {
+      field = field.withNullable(true);
+    } else if ("REQUIRED".equals(column.getMode())) {
+      field = field.withNullable(false);
+    } else if ("REPEATED".equals(column.getMode())) {
+      field = Field.of(name, FieldType.array(fieldType));
```

Comment:
How's nullability on element type of ARRAY. Say ARRAY<INT64> but it could be (null, null, 12, 13, null...)

Reply:
BQ does not allow nulls in arrays: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#array-type

JSON Arrays can potentially include `nulls`, not sure whether we should allow this or not